lib/scholar/metrics/distance.ex

```diff
@@ -468,20 +521,29 @@ defmodule Scholar.Metrics.Distance do
         f32
         0.75
       >
+
+      iex> x = Nx.tensor([[6, 2, 9], [2, 5, 3]])
+      iex> y = Nx.tensor([[8, 3, 1]])
+      iex> weights = Nx.tensor([1, 0.5, 0.5])
+      iex> Scholar.Metrics.Distance.hamming(x, y, weights, axes: [1])
+      #Nx.Tensor<
+        f32[2]
+        [1.0, 1.0]
+      >
   """
   deftransform hamming(x, y, w, opts) when is_list(opts) do
     NimbleOptions.validate!(opts, @general_schema)
     hamming_weighted(x, y, w, opts)
   end
 
   defnp hamming_unweighted(x, y, opts) do
-    assert_same_shape!(x, y)
+    valid_broadcast?(Nx.rank(x), Nx.shape(x), Nx.shape(y))
     result_type = Nx.Type.to_floating(Nx.Type.merge(Nx.type(x), Nx.type(y)))
     Nx.mean(x != y, axes: opts[:axes]) |> Nx.as_type(result_type)
   end
 
   defnp hamming_weighted(x, y, w, opts) do
-    assert_same_shape!(x, y)
+    valid_broadcast?(Nx.rank(x), Nx.shape(x), Nx.shape(y))
```

Comment:
Perhaps we should instead check if they have the same `rank`? And then, if broadcast fails, it will naturally fail in a later operation?

Otherwise, if we want to keep the current code, it doesn't currently raise if the broadcast shape do not match, so this is not checking much (and it also fails with an unrelated exception if they don't have the same rank).

Reply:
I'll add a check on rank and raise an error in both cases in `valid_broadcast?`

Reply:
I added some changes, WDYT @josevalim?